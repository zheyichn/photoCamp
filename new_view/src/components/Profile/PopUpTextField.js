import TextField from '@mui/material/TextField';
// import { useEffect, useState } from 'react';

export default function PopUpTextField(props) {

    const {field, postInfo, defaultValue} = props;

    let textValue = defaultValue;
    
    if (field === 'imageSrc') {
        postInfo.current.imageSrc = textValue;
    }

    if (field === 'postCaption') {
        postInfo.current.postCaption = textValue;
    }

    const changeRef = (e) => {
        textValue = e.target.value;
        if (field === 'imageSrc') {
            postInfo.current.imageSrc = textValue;
        }
        if (field === 'postCaption') {
            postInfo.current.postCaption = textValue;
        }
    }

    return(
        <TextField fullWidth onChange={changeRef} defaultValue={defaultValue || ''} />
    );
}