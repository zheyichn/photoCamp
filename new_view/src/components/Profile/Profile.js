import React, { useState, useEffect, useRef } from "react";
import { s3Operation } from "../../utility/s3Operation";
import {
  getUserAPI,
  updateUserProfile,
} from "../../api/mock_api_following_post_follower";

export default function Profile() {
  const [followersNum, setfollowersNum] = useState(0);
  const [followingNum, setfollowingNum] = useState(0);
  const [postsNum, setpostsNum] = useState(0);
  const [userInfo, setuserInfo] = useState({});

  const [avatar, setAvatar] = useState([
    "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png",
  ]);

  const handleFollowerClick = () => (window.location.pathname = "/follower");
  const handleFollowingClick = () => (window.location.pathname = "/following");

  const hiddenFileInput = useRef(null);
  const handleClick = () => {
    hiddenFileInput.current.click();
  };

  async function handleChange(event) {
    const imgUrl = await s3Operation(event.target.files[0]);
    setAvatar(imgUrl);
    await updateUserProfile(sessionStorage.userName, imgUrl);
    sessionStorage.setItem("userProfile", imgUrl);
  }

  const getFollowing = async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    const newFollowingNum = userResponse.data.followings.length;
    setfollowingNum(newFollowingNum);
  };

  const getFollower = async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    const newFollowersNum = userResponse.data.followers.length;
    setfollowersNum(newFollowersNum);
  };

  const getPosts = async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    const newPostsNum = userResponse.data.posts.length;
    setpostsNum(newPostsNum);
  };

  const getUserInfo = async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    setuserInfo(userResponse.data);
    setAvatar(userResponse.data.profile);
  };

  useEffect(() => {
    getUserInfo();
    getFollowing();
    getFollower();
    getPosts();
  }, []);

  return (
    <div className="profile-header">
      <div className="container">
        <div className="profile">
          <div className="profile-photo">
            <img src={avatar} alt="" />
          </div>

          <div className="profile-user-settings">
            <h1 className="profile-user-name">{userInfo.userName}</h1>

            <button className="edit-profile-btn" onClick={handleClick}>
              Change Photo
            </button>

            <input
              type="file"
              id="input-btn"
              ref={hiddenFileInput}
              onChange={handleChange}
              hidden
            />
          </div>

          <div className="profile-stats">
            <ul>
              <li>
                <span className="profile-info-follow">{postsNum}</span> posts
              </li>
              <li>
                <button
                  className="btn profile-info-follow"
                  onClick={handleFollowerClick}
                >
                  {followersNum + " followers"}
                </button>{" "}
              </li>
              <li>
                <button
                  className="btn profile-info-follow"
                  onClick={handleFollowingClick}
                >
                  {followingNum + " following"}
                </button>
              </li>
            </ul>
          </div>

          <div className="profile-bio">
            <div>
              <p className="bio">{userInfo.bio}</p>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}
