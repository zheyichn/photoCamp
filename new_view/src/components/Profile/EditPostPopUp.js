import CardMedia from '@mui/material/CardMedia';
import Dialog from "@mui/material/Dialog";
import DialogTitle from "@mui/material/DialogTitle";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import Grid from "@mui/material/Grid";
import PopUpTextField from "./PopUpTextField";
import { useEffect, useRef, useState } from "react";
import { putPostEdit, deletePost } from "../../api/mock_api_j";
import VisibilityIcon from "@mui/icons-material/Visibility";
import { unhidePost } from "../../api/mock_api_k";

export default function EditPostPopUp(props) {
  const { open, onClose, currPost } = props;
  const [isHidden, setIsHidden] = useState(
    currPost.hiddenFrom
      ? currPost.hiddenFrom.includes(sessionStorage.getItem("userName"))
      : false
  );

  useEffect(() => {
    setIsHidden(
      currPost.hiddenFrom
        ? currPost.hiddenFrom.includes(sessionStorage.getItem("userName"))
        : false
    );
  }, [currPost]);

  // 1. Initialize a Ref variable that stores the key information
  const postInfo = useRef({ imageSrc: "", postCaption: "" });

  // 2. Initialize a function that logs the ref variable on save
  // to be changed to api requests later
  // 3. Pass down to pop up text field
  const onSave = () => {
    let amendedPost = currPost; // does NOT create deep copy but changed the var name for readability
    amendedPost.caption = postInfo.current.postCaption;
    // amendedPost.postContent = postInfo.current.imageSrc;
    putPostEdit(amendedPost);
    onClose();
  };

  const onDelete = () => {
    console.log(currPost._id);
    deletePost(currPost._id);
    onClose();
    setTimeout(() => {
      window.location.reload();
    }, 200);
  };

  const onUnHide = async () => {
    try {
      await unhidePost(currPost._id, sessionStorage.getItem("userName"));
      setIsHidden(false);
      setTimeout(() => {
        window.location.reload();
      }, 100);
    } catch (err) {
      console.err(err);
    }
  };

  return (
    <Dialog open={open} onClose={onClose}>
      <Grid container justifyContent="center" alignItems="center">
        <Grid item>
          <DialogTitle style={{ padding: 16, textAlign: "center" }}>
            Edit My Post
          </DialogTitle>
        </Grid>
        {isHidden ? (
          <Grid item>
            <VisibilityIcon mt={2} onClick={onUnHide}></VisibilityIcon>
          </Grid>
        ) : null}
      </Grid>
      { currPost.isImage? 
      <img
        srcSet={`${currPost.postContent}?w=500&h=500&fit=crop&auto=format&dpr=2`}
        alt={currPost.title}
        loading="lazy"
      /> 
      : 
      <CardMedia
        component="video"
        image={`${currPost.postContent}`}
        controls
        alt="video"
        autoPlay
        muted
        loop
      /> }
      <Grid container flex-direction="row" padding={1}>
        <Grid item xs={3}>
          <Typography style={{ padding: 16 }}>Post caption:</Typography>
        </Grid>
        <Grid item xs={9}>
          <PopUpTextField
            field="postCaption"
            defaultValue={currPost.caption}
            postInfo={postInfo}
          />
        </Grid>
        <Grid item xs={4}>
          <Button
            data-testid="save-btn"
            onClick={onSave}
            style={{ padding: 16 }}
          >
            Save Changes
          </Button>
        </Grid>
        <Grid item xs={4}>
          <Button onClick={onClose} style={{ padding: 16 }}>
            Cancel Changes
          </Button>
        </Grid>
        <Grid item xs={4}>
          <Button
            variant="outlined"
            color="error"
            onClick={onDelete}
            style={{ padding: 16 }}
          >
            Delete Post
          </Button>
        </Grid>
      </Grid>
    </Dialog>
  );
}
