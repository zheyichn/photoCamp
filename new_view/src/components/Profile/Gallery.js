import { useState, useEffect, useCallback } from "react";
import ImageList from "@mui/material/ImageList";
import { CardMedia } from "@mui/material";
import ImageListItem from "@mui/material/ImageListItem";
import {
  getTagsAPI,
  getUserAPI,
  getPostByIdAPI,
} from "../../api/mock_api_following_post_follower";
import EditPostPopUp from "./EditPostPopUp";
import { Box, Tab, Tabs } from "@mui/material";
import { isImgUrl } from "../../utility/isImageUrl";

export default function Gallery() {
  const [posts, setPosts] = useState([]);
  const [tagPosts, setTagPosts] = useState([]);
  const [, setWhichPost] = useState([]);
  // states to determine whether edit post pop-up is open and what is the current post being edited
  const [editOpen, setEditOpen] = useState(false);
  const [currPost, setCurrPost] = useState("");
  const [tabIndex, setTabIndex] = useState(0);

  const judgeIfImage = async (url) => {
    const res = await isImgUrl(url);
    return res;
  };

  const handleTabChange = (event, newTabIndex) => {
    setTabIndex(newTabIndex);
    if (tabIndex === 0) {
      setWhichPost(tagPosts);
    } else {
      setWhichPost(posts);
    }
  };

  // handle the opening of edit post pop up
  const handleClickOpen = (post) => {
    setEditOpen(true);
    setCurrPost(post);
  };

  // handle the closing of edit post pop up
  const handleClickClose = () => {
    setEditOpen(false);
    setCurrPost("");
  };

  // get one post
  const getOnePost = useCallback(async (postId) => {
    let post = await getPostByIdAPI(postId);
    post = post.data[0];
    const isImage = await judgeIfImage(post.postContent);
    post["isImage"] = isImage;
    return post;
  }, []);

  // get all the posts associated with this user from api
  const getPostContent = useCallback(async () => {
    const userResponse = await getUserAPI(sessionStorage.userName);
    const postIds = userResponse.data.posts;
    const postsContent = await Promise.all(postIds.map(getOnePost));
    setPosts(postsContent);
  }, [getOnePost]);


  const judgeImageHelper = useCallback(async(post) => {
    const isImage = await judgeIfImage(post.postContent);
    post["isImage"] = isImage;
    return post;
  }, []);
  
  // get tab content associated with the current user
  // and judge whether each post is an image
  const getTagContent = useCallback(async () => {
    const tagResponse = await getTagsAPI(sessionStorage.userName);
    const taggedPosts = tagResponse.data;
    const judgedTaggedPosts = await Promise.all(taggedPosts.map(judgeImageHelper));
    setTagPosts(judgedTaggedPosts);
  }, [judgeImageHelper]);



  // call getPostContent()
  // this should be called initially, AND whenever the pop up opens / closes
  // allow a timeout of 100ms - sometimes we perform a POST request to back end and we are fetching the data again
  useEffect(() => {
    getPostContent();
    getTagContent();
  }, [getPostContent, getTagContent]);

  useEffect(() => {
    setTimeout(() => getPostContent(), 100);
    setTimeout(() => getTagContent(), 100);
    // setTimeout(() => getTagContent(), 100);
  }, [editOpen, getPostContent, getTagContent]);

  return (
    <main>
      <Box>
        <Tabs value={tabIndex} onChange={handleTabChange} centered>
          <Tab label="Posts" />
          <Tab label="Tagged" />
        </Tabs>
      </Box>

      <div className="container">
        {tabIndex ? (
          <div className="gallery">
            <ImageList
              data-testid="img-list"
              sx={{ width: 1000, height: 1000 }}
              cols={3}
              rowHeight={333}
            >
              {tagPosts.map((item) => (
                <ImageListItem key={item._id}>
                  {/* Note: use this when implenting backend for tagging!*/}
                  {item.isImage ? (
                    <CardMedia
                      data-testid="test-img"
                      component="img"
                      height="300px"
                      src={item.postContent}
                      alt="Post Picture"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  ) : (
                    <CardMedia
                      data-testid="test-img"
                      component="video"
                      height="300px"
                      src={item.postContent}
                      autoPlay
                      controls
                      muted
                      alt="Post Video"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  )}
                </ImageListItem>
              ))}
            </ImageList>
          </div>
        ) : (
          <div className="gallery">
            <ImageList
              data-testid="img-list"
              sx={{ width: 1000, height: 1000 }}
              cols={3}
              rowHeight={333}
            >
              {posts.map((item) => (
                <ImageListItem key={item._id}>
                  {item.isImage ? (
                    <CardMedia
                      data-testid="test-img"
                      component="img"
                      height="300px"
                      src={item.postContent}
                      alt="Post Picture"
                      onClick={() => handleClickOpen(item)}
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  ) : (
                    <CardMedia
                      data-testid="test-img"
                      component="video"
                      height="300px"
                      src={item.postContent}
                      autoPlay
                      controls
                      muted
                      onClick={() => handleClickOpen(item)}
                      alt="Post Video"
                      sx={{
                        filter: item.hiddenFrom.includes(
                          sessionStorage.getItem("userName")
                        )
                          ? "blur(3px)"
                          : null,
                      }}
                    ></CardMedia>
                  )}
                </ImageListItem>
              ))}
            </ImageList>
          </div>
      )}

        <EditPostPopUp
          open={editOpen}
          onClose={handleClickClose}
          currPost={currPost}
        />
      </div>
    </main>
  );
}
