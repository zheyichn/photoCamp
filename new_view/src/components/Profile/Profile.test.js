import { cleanup, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import Profile from "./Profile";

// afterEach function runs after each test suite is executed
afterEach(() => {
  cleanup();
});

describe("Profile", () => {
  it("Page renders correctly", () => {
    const tree = renderer.create(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  //Test2:
  test("profile name displayed", () => {
    render(<Profile />);
    const ele = screen.getByText("posts");
    expect(ele).toBeInTheDocument();
    expect(ele).toHaveTextContent("posts");
  });
});
