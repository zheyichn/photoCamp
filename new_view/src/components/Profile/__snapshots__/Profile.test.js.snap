// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Page renders correctly 1`] = `
<div
  className="profile-header"
>
  <div
    className="container"
  >
    <div
      className="profile"
    >
      <div
        className="profile-photo"
      >
        <img
          alt=""
          src={
            [
              "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png",
            ]
          }
        />
      </div>
      <div
        className="profile-user-settings"
      >
        <h1
          className="profile-user-name"
        />
        <button
          className="edit-profile-btn"
          onClick={[Function]}
        >
          Edit Profile
        </button>
        <button
          className="edit-profile-btn"
          onClick={[Function]}
        >
          Change Photo
        </button>
        <input
          hidden={true}
          id="input-btn"
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className="profile-stats"
      >
        <ul>
          <li>
            <span
              className="profile-info-follow"
            >
              0
            </span>
             posts
          </li>
          <li>
            <button
              className="btn profile-info-follow"
              onClick={[Function]}
            >
              0 followers
            </button>
             
          </li>
          <li>
            <button
              className="btn profile-info-follow"
              onClick={[Function]}
            >
              0 following
            </button>
          </li>
        </ul>
      </div>
      <div
        className="profile-bio"
      >
        <div>
          <p
            className="bio"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
